EventFilter/EcalRawToDigi/plugins/alpaka/EcalRawToDigiPortable.cc

```diff
@@ -0,0 +1,146 @@
+
+#include "DeclsForKernels.h"
+#include "UnpackPortable.h"
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+
+  class EcalRawToDigiPortable : public stream::EDProducer<> {
+  public:
+    explicit EcalRawToDigiPortable(edm::ParameterSet const& ps);
+    ~EcalRawToDigiPortable() override = default;
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+    void produce(device::Event&, device::EventSetup const&) override;
+
+  private:
+    edm::EDGetTokenT<FEDRawDataCollection> rawDataToken_;
+    using OutputProduct = EcalDigiDeviceCollection;
+    device::EDPutToken<OutputProduct> digisDevEBToken_;
+    device::EDPutToken<OutputProduct> digisDevEEToken_;
+    device::ESGetToken<EcalElectronicsMappingDevice, EcalMappingElectronicsRcd> eMappingToken_;
+
+    std::vector<int> fedsToUnpack_;
+
+    ecal::raw::ConfigurationParameters config_;
+  };
+
+  void EcalRawToDigiPortable::fillDescriptions(edm::ConfigurationDescriptions& confDesc) {
+    edm::ParameterSetDescription desc;
+
+    desc.add<edm::InputTag>("InputLabel", edm::InputTag("rawDataCollector"));
+    std::vector<int> feds(54);
+    for (uint32_t i = 0; i < 54; ++i)
+      feds[i] = i + 601;
+    desc.add<std::vector<int>>("FEDs", feds);
+    desc.add<uint32_t>("maxChannelsEB", 61200);
+    desc.add<uint32_t>("maxChannelsEE", 14648);
+    desc.add<std::string>("digisLabelEB", "ebDigis");
+    desc.add<std::string>("digisLabelEE", "eeDigis");
+
+    confDesc.addWithDefaultLabel(desc);
+  }
+
+  EcalRawToDigiPortable::EcalRawToDigiPortable(const edm::ParameterSet& ps)
+      : rawDataToken_{consumes<FEDRawDataCollection>(ps.getParameter<edm::InputTag>("InputLabel"))},
+        digisDevEBToken_{produces(ps.getParameter<std::string>("digisLabelEB"))},
+        digisDevEEToken_{produces(ps.getParameter<std::string>("digisLabelEE"))},
+        eMappingToken_{esConsumes()},
+        fedsToUnpack_{ps.getParameter<std::vector<int>>("FEDs")} {
+    config_.maxChannelsEB = ps.getParameter<uint32_t>("maxChannelsEB");
+    config_.maxChannelsEE = ps.getParameter<uint32_t>("maxChannelsEE");
+  }
+
+  void EcalRawToDigiPortable::produce(device::Event& event, device::EventSetup const& setup) {
+    // conditions
+    auto const& eMappingProduct = setup.getData(eMappingToken_);
+
+    // event data
+    const auto rawDataHandle = event.getHandle(rawDataToken_);
+
+    // make a first iteration over the FEDs to compute the total buffer size
+    uint32_t size = 0;
+    uint32_t feds = 0;
+    for (auto const& fed : fedsToUnpack_) {
+      auto const& data = rawDataHandle->FEDData(fed);
+      auto const nbytes = data.size();
+
+      // skip empty FEDs
+      if (nbytes < globalFieds::EMPTYEVENTSIZE)
+        continue;
+
+      size += nbytes;
+      ++feds;
+    }
+
+    auto& queue = event.queue();
+
+    // input host buffers
+    ecal::raw::InputDataHost inputHost(queue, size, feds);
+
+    // input device buffers
+    ecal::raw::InputDataDevice inputDevice(queue, size, feds);
```

Comment:
`inputDevice` does not seem to be used in this function.
Instead of allocating it here and passing it by reference, can you move it directly inside ` ecal::raw::unpackRaw()` ?

Reply:
Ah no not needed. I could use `fedCounter' and `currentCummOffset` instead. They contain the same values.